# Data Preparation

In [1]:
# Import necessary libraries
import pandas as pd  # For data manipulation and analysis
import numpy as np   # For numerical operations
import seaborn as sns  # For statistical data visualization
import matplotlib.pyplot as plt  # For creating static, animated, and interactive visualizations
import scipy.stats as stats  # For statistical functions

In [2]:
# Import feature selection and preprocessing modules from scikit-learn
from sklearn.feature_selection import SelectKBest, f_classif  # For selecting the best features based on ANOVA F-statistic
from sklearn.decomposition import PCA  # For dimensionality reduction using Principal Component Analysis
from sklearn.preprocessing import StandardScaler  # For standardizing features by removing the mean and scaling to unit variance
from sklearn.preprocessing import OneHotEncoder  # For converting categorical variables into a form that can be provided to ML algorithms
from sklearn.preprocessing import LabelEncoder  # For encoding target labels with a value between 0 and n_classes-1
from sklearn.preprocessing import OrdinalEncoder  # For encoding ordinal categorical features with integer values

In [3]:
import os  # For interacting with the operating system
import sys  # For system-specific parameters and functions

# Add the 'scripts' directory to the Python path for module imports
sys.path.append(os.path.abspath(os.path.join('..', 'scripts')))

# Set maximum number of rows and columns to display in pandas DataFrames
pd.set_option('display.max_columns', 200)  # Maximum columns to display
pd.set_option('display.max_rows', 200)     # Maximum rows to display

In [4]:
# Read the dataset
df = pd.read_csv('../data/cleaned_data.csv', low_memory=False, index_col=False)

In [5]:
# Explore the first ten rows
df.head(10)

Unnamed: 0  UnderwrittenCoverID  PolicyID     TransactionMonth  \
0           0               145249     12827  2015-03-01 00:00:00   
1           1               145249     12827  2015-05-01 00:00:00   
2           2               145249     12827  2015-07-01 00:00:00   
3           3               145255     12827  2015-05-01 00:00:00   
4           4               145255     12827  2015-07-01 00:00:00   
5           5               145247     12827  2015-01-01 00:00:00   
6           6               145247     12827  2015-04-01 00:00:00   
7           7               145247     12827  2015-06-01 00:00:00   
8           8               145247     12827  2015-08-01 00:00:00   
9           9               145245     12827  2015-03-01 00:00:00   

   IsVATRegistered Citizenship          LegalType Title Language  \
0             True              Close Corporation    Mr  English   
1             True              Close Corporation    Mr  English   
2             True              Close Corporation    Mr  English   
3             True              Close Corporation    Mr  English   
4             True              Close Corporation    Mr  English   
5             True              Close Corporation    Mr  English   
6             True              Close Corporation    Mr  English   
7             True              Close Corporation    Mr  English   
8             True              Close Corporation    Mr  English   
9             True              Close Corporation    Mr  English   

                  Bank      AccountType  MaritalStatus         Gender  \
0  First National Bank  Current account  Not specified  Not specified   
1  First National Bank  Current account  Not specified  Not specified   
2  First National Bank  Current account  Not specified  Not specified   
3  First National Bank  Current account  Not specified  Not specified   
4  First National Bank  Current account  Not specified  Not specified   
5  First National Bank  Current account  Not specified  Not specified   
6  First National Bank  Current account  Not specified  Not specified   
7  First National Bank  Current account  Not specified  Not specified   
8  First National Bank  Current account  Not specified  Not specified   
9  First National Bank  Current account  Not specified  Not specified   

        Country Province  PostalCode MainCrestaZone SubCrestaZone  \
0  South Africa  Gauteng        1459      Rand East     Rand East   
1  South Africa  Gauteng        1459      Rand East     Rand East   
2  South Africa  Gauteng        1459      Rand East     Rand East   
3  South Africa  Gauteng        1459      Rand East     Rand East   
4  South Africa  Gauteng        1459      Rand East     Rand East   
5  South Africa  Gauteng        1459      Rand East     Rand East   
6  South Africa  Gauteng        1459      Rand East     Rand East   
7  South Africa  Gauteng        1459      Rand East     Rand East   
8  South Africa  Gauteng        1459      Rand East     Rand East   
9  South Africa  Gauteng        1459      Rand East     Rand East   

           ItemType      mmcode        VehicleType  RegistrationYear  \
0  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
1  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
2  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
3  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
4  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
5  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
6  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
7  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
8  Mobility - Motor  44069150.0  Passenger Vehicle              2004   
9  Mobility - Motor  44069150.0  Passenger Vehicle              2004   

            make  Model  Cylinders  cubiccapacity  kilowatts bodytype  \
0  MERCEDES-BENZ  E 240        6.0         2597.0      130.0    

In [6]:
# Check for missing values in each column of the DataFrame
missing_counts = df.isnull().sum()

# Identify columns that have more than 1 missing value
columns_with_missing = missing_counts[missing_counts > 1].index
print(f'Columns with more than 1 missing value:\n{columns_with_missing}')

# Drop these columns from the DataFrame
df = df.drop(columns=columns_with_missing)

# Verify the result by printing the remaining columns
print(f'Columns remaining after dropping:\n{df.columns}')

Columns with more than 1 missing value:
Index(['VehicleIntroDate'], dtype='object')
Columns remaining after dropping:
Index(['Unnamed: 0', 'UnderwrittenCoverID', 'PolicyID', 'TransactionMonth',
       'IsVATRegistered', 'Citizenship', 'LegalType', 'Title', 'Language',
       'Bank', 'AccountType', 'MaritalStatus', 'Gender', 'Country', 'Province',
       'PostalCode', 'MainCrestaZone', 'SubCrestaZone', 'ItemType', 'mmcode',
       'VehicleType', 'RegistrationYear', 'make', 'Model', 'Cylinders',
       'cubiccapacity', 'kilowatts', 'bodytype', 'NumberOfDoors',
       'AlarmImmobiliser', 'TrackingDevice', 'CapitalOutstanding',
       'NewVehicle', 'SumInsured', 'TermFrequency', 'CalculatedPremiumPerTerm',
       'ExcessSelected', 'CoverCategory', 'CoverType', 'CoverGroup', 'Section',
       'Product', 'StatutoryClass', 'StatutoryRiskType', 'TotalPremium',
       'TotalClaims'],
      dtype='object')


In [9]:
# Get categorical columns from the DataFrame
categorical_columns = df.select_dtypes(include='object').columns

# Create a summary DataFrame to hold information about categorical columns
summary_df = pd.DataFrame({
    'Column': categorical_columns,  # Column names
    'DataType': [df[col].dtype for col in categorical_columns],  # Data types of each column
    'NumUniqueValues': [df[col].nunique() for col in categorical_columns]  # Number of unique values in each column
})

# Print the summary DataFrame
print(f'Display the DataFrame with the summary:')
summary_df  # This will display the DataFrame with the summary

Display the DataFrame with the summary:


Column DataType  NumUniqueValues
0     TransactionMonth   object               23
1          Citizenship   object                4
2            LegalType   object                6
3                Title   object                5
4             Language   object                1
5                 Bank   object               11
6          AccountType   object                3
7        MaritalStatus   object                3
8               Gender   object                3
9              Country   object                1
10            Province   object                9
11      MainCrestaZone   object               16
12       SubCrestaZone   object               45
13            ItemType   object                1
14         VehicleType   object                5
15                make   object               46
16               Model   object              411
17            bodytype   object               13
18    AlarmImmobiliser   object                2
19      TrackingDevice   object                2
20  CapitalOutstanding   object              659
21          NewVehicle   object                2
22       TermFrequency   object                2
23      ExcessSelected   object               13
24       CoverCategory   object               28
25           CoverType   object               22
26          CoverGroup   object               14
27             Section   object                5
28             Product   object                4
29      StatutoryClass   object                1
30   StatutoryRiskType   object                1

# Key Insights from the Dataset Summary

## Overview
This document summarizes key findings from the dataset, highlighting columns to consider for removal or transformation to enhance model performance.

## Insights

- **Low Variability Columns**: 
  - Columns such as **Language**, **Country**, **ItemType**, **StatutoryClass**, and **StatutoryRiskType** exhibit minimal variability and should be dropped to streamline the dataset.

- **High Cardinality Column**:
  - The **Model** column has 411 unique values, potentially adding complexity. Consider reducing its cardinality through grouping or applying target encoding.

- **Binary Columns**: 
  - Columns like **AlarmImmobiliser**, **TrackingDevice**, and **NewVehicle** contain only 2 unique values, which may not provide significant predictive power.

- **Date Conversion**:
  - **TransactionMonth** should be converted to a date format. Extracting features such as year, month, and quarter may enhance the model.

- **Continuous Feature**:
  - **CapitalOutstanding** should be treated as a numeric feature, allowing it to be utilized as a continuous variable in modeling.

## Next Steps
- Remove low variability columns.
- Apply transformations to high cardinality and date-related columns.
- Evaluate the impact of binary columns on model performance.